### PR TITLE
Implementing ASE without the motion-capture component 

### PR DESCRIPTION
**Pull request by dtch1997**

### CODE REVIEW

**Comment by dtch1997** [comment]
Some TODOs:
- Provide encoder log probability to the policy as its reward. 
- Use `torch.sum` instead of `torch.mean` to compute encoder loss.

**Comment by dtch1997** [comment]
Our first goal here is to be able to reproduce one of the results from DADS:  https://arxiv.org/pdf/1907.01657.pdf
In Fig 4, diverse skill embeddings on the Ant environment learned a controller that can move in different directions

**Comment by dtch1997** [comment]
Next TODO: Write the eval script that loads checkpoint and opens the IsaacGym viewer

**Comment by dtch1997** [comment]
One key insight from today: 

In skill learning, in addition to diversity and controllability, **utility** of the skill for downstream behaviours is important. 
The DADS framework hopes to achieve utility through maintaining a skill set that explores a large part of state space, but some parts of state space are not inherently interesting to explore in. E.g. the Ant has 111 observations, but arguably only the body velocity and angular velocity are really interesting to explore in. 

In DADS, utility was achieved (in Ant) by restricting the observation of the skill dynamics module to the (x,y) position of the body. I.e. all skills will be evaluated based only on how well they achieve predictable changes in (x,y) position, which regulates the learning towards directional locomotion. 

In ASE, utility is achieved in a data-driven manner by having a pre-existing library of several stock behaviours which the policy must learn to reproduce (through adversarial imitation). 

I think utility is best achieved in a task-oriented way, e.g. through curriculum learning. This might seem odd because the point of unsupervised RL is to avoid task engineering; however we want to only specify tasks in a simple way with minimal reward shaping, and my theory is that a sufficiently diverse suite of tasks (and initializations!) suffices as an inductive bias on skills. 

Another hypothesis is to make use of "physical priors" on skills. Similar to the 'robotic priors' paper from Oliver Brock here, we can hypothesise that useful motor skills satisfy certain basic physical criteria. For example, energy efficiency. 

Another possibility is to try to use representation learning to limit the  exploration of the state space to those dimensions that are truly interesting. Put another way, we want to explore skills that enable transitions in the low-dimensional 'meaningful manifold' and not the high-dimensional orthogonal directions that largely don't matter for downstream tasks.

**Comment by dtch1997** [comment]
To check the correctness of implementation until now, we need to figure out whether the learned skill embeddings satisfy the following two criteria: 

1. Diversity - do different skill embeddings indeed map to different types of skill? 
2. Predictability - given a single skill embedding, is the following sequence of actions truly predictable? 

As a sanity check for now, we can do this by inspection with manually-designed metrics in simple environments, but we should be careful not to invest too much time in that and instead think about more general metrics that apply universally.

**Comment by dtch1997** [comment]
From visual inspection of learned checkpoints, it seems that the current implementation fails to create similar skills as DADS. I will try again using an approach more similar to DADS and hope to reproduce some of the figures in their paper.

**Comment by dtch1997** [comment]
New idea: 

## Skill Learning with Skill Priors

As we have seen from previous papers, task-specific prior knowledge aids immensely in the learning of effective and **useful** skills for downstream tasks. Typically, such prior knowledge is encoded in the form of task rewards, but those can be difficult to design. Instead we propose skill priors which are inductive biases on what constitutes an effective and useful skills. 

Previously, priors on skills have been proposed, such as:
- The (x,y) prior for locomotion (DADS) 
- Adversarial motion priors (AMP, ASE)

We can think of more priors that might be satisfied by 'useful' skills: 
- High predictability of certain state dimensions (a generalization of x,y prior) 
- Energy-efficiency

There are typically also algorithmic priors on skills relating to the way in which we discover those skills:
- Empowerment (DADS, DIAYN) 
This is not the subject of our investigation

**Comment by dtch1997** [comment]
Next TODO: 

Test unsupervised RL method on simple benchmark like 2D point navigation and figure out appropriate metrics.
